From 2dd8afc19613ca3f50c578870b6d3db8006ccb8c Mon Sep 17 00:00:00 2001
From: Behan Webster <behanw@converseincode.com>
Date: Wed, 14 Nov 2012 12:55:28 +0100
Subject: [PATCH] x86: LLVMLinux: Reimplement current_stack_pointer
 without register usage.
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

Use asm to make the globally named register work again for gcc and clang.
Much more efficient than copying the stack pointer to a variable and back again.

Signed-off-by: Behan Webster <behanw@converseincode.com>
---

--- linux.orig/arch/x86/include/asm/thread_info.h
+++ linux/arch/x86/include/asm/thread_info.h
@@ -163,10 +163,10 @@
  */
 #ifndef __ASSEMBLY__
 
-#define current_stack_pointer ({		\
-	unsigned long sp;			\
-	asm("mov %%esp,%0" : "=g" (sp));	\
-	sp;					\
+#define current_stack_pointer ({			\
+	register unsigned long sp asm("esp") __used;	\
+	asm("" : "=r" (sp));				\
+	sp;						\
 })
 
 /* how to get the thread information struct from C */
